Let recorder deal with event names longer than 32 chars

lmagyar opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The `recorder` component's `events.event_type` column is too small for some events, causing SQL errors in the log. VARCHAR(32) increased to VARCHAR(64),  though it is still smaller than the similar `states.entity_id(255)`.

See #38917

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #38917
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

lmagyar committed:
Deal with event names longer than 32 chars

lmagyar commented:
I'm new to HA, I have no proper dev environment yet, so my test was quite simple: copied the new recorder code into the running core container, restarted it, and it worked, the DB schema changed, `schema_changes` updated, see the log below (it implicitly tested changes in #47026 also).

Previously I've modified my SQL schema (varchar 32->64) directly in my MariaDB, it was running for a few weeks without problems.

My config: MariaDB 10.5.8, with Aria storage engine on tmpfs (in-memory DB). I'm using MariaDB 10.5.8 with Alpine Linux 3.13 because max key length is 2000 bytes in this newer version, and `entity_id(255)` was too big for 1000 bytes hardcoded in MariaDB 10.4.17.

```
2021-03-11 10:42:05 WARNING (Recorder) [homeassistant.components.recorder.migration] Database is about to upgrade. Schema version: 11
2021-03-11 10:42:05 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_data in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-11 10:42:14 WARNING (MainThread) [homeassistant.setup] Setup of recorder is taking over 10 seconds.
2021-03-11 10:42:27 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns attributes in table states. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-11 10:42:54 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-11 10:43:02 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: recorder
```

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -363,6 +363,7 @@ def _apply_update(engine, new_version, old_version):
         if engine.dialect.name == "mysql":
             _modify_columns(engine, "events", ["event_data LONGTEXT"])
             _modify_columns(engine, "states", ["attributes LONGTEXT"])
+        _modify_columns(engine, "events", ["event_type VARCHAR(64)"])
```
I'm pretty sure this won't work for sqlite

bdraco replied:
Since copying all the data to a new table is likely going to be a very long running operation, I think we can exclude sqlite here.

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -363,6 +363,7 @@ def _apply_update(engine, new_version, old_version):
         if engine.dialect.name == "mysql":
             _modify_columns(engine, "events", ["event_data LONGTEXT"])
             _modify_columns(engine, "states", ["attributes LONGTEXT"])
+        _modify_columns(engine, "events", ["event_type VARCHAR(64)"])
```
```suggestion
        if instance.engine.dialect.name == "sqlite"
            # sqlite cannot alter columns
            _modify_columns(engine, "events", ["event_type VARCHAR(64)"])
```

bdraco replied:
This fails on postgresql
```
2021-03-11 08:33:13 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and sl
2021-03-11 08:33:13 ERROR (Recorder) [homeassistant.components.recorder.util] Error executing query: (psycopg2.errors.SyntaxError) syntax error at or near "MODIFY"
LINE 1: ALTER TABLE events MODIFY event_type VARCHAR(64)
                           ^

[SQL: ALTER TABLE events MODIFY event_type VARCHAR(64)]
(Background on this error at: http://sqlalche.me/e/13/f405)
2021-03-11 08:33:13 ERROR (Recorder) [homeassistant.components.recorder] Error during connection setup to postgresql://postgres:homeassistant@77b2833f-timescaledb/homeassistant: (psycopg2.erro
LINE 1: ALTER TABLE events MODIFY event_type VARCHAR(64)
                           ^

[SQL: ALTER TABLE events MODIFY event_type VARCHAR(64)]
(Background on this error at: http://sqlalche.me/e/13/f405) (retrying in 3 seconds)
```

bdraco replied:
https://dbfiddle.uk/?rdbms=postgres_9.5&fiddle=40f8bd4ca9c9d4cb599183668478b085

bdraco replied:
https://dbfiddle.uk/?rdbms=postgres_9.5&fiddle=b694eb2bec4e4ff56595acc83301aea3

bdraco replied:
I think it makes sense to only do it for mysql.  Anyone else who wants the larger size can recreate the database or do it manually.

bdraco commented:
Testing matrix:

- [ ] sqlite3
- [ ] mysql
- [ ] postgresql

lmagyar committed:
Handle sqlite, postgresql, mssql

skip sqlite
add special handling for postgresql and mssql
default works for mysql and mariadb (and should work for oracle)

lmagyar reviewed (commented):
I will make it work on postgresql, maybe later there will be a real need to modify something.

lmagyar commented:
I made modifications:
- skip [sqlite](https://sqlite.org/lang_altertable.html), it can't handle it, we should execute it in several steps
- added special handling to make it work on: [postgresql](https://www.postgresql.org/docs/9.1/sql-altertable.html) and [mssql](https://docs.microsoft.com/en-us/sql/t-sql/statements/alter-table-transact-sql?view=sql-server-ver15)
- default works for [mysql](https://dev.mysql.com/doc/refman/8.0/en/alter-table.html)/mariadb (and should work for [oracle](https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_3001.htm#i2103924))

If somebody accidentally tries to modify sqlite later, a warning will be written in the log, instead of a silent skip!

Testing matrix:
- [ ] sqlite - excluded
- [x] mysql/mariadb (mariadb mariadb 10.5.8)
- [x] postgresql (postgresql 13)
- [x] mssql (mssql express 2014)
- [ ] oracle - not tested, but should work with default mysql syntax

## Test results

### sqlite

Not tested, it should skip the modification and increase schema version.

### mysql/mariadb

```
2021-03-12 19:09:52 WARNING (Recorder) [homeassistant.components.recorder.migration] Database is about to upgrade. Schema version: 11
2021-03-12 19:09:52 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_data in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-12 19:10:02 WARNING (MainThread) [homeassistant.setup] Setup of recorder is taking over 10 seconds.
2021-03-12 19:10:15 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns attributes in table states. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-12 19:10:42 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-03-12 19:10:50 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: recorder
```

### postgresql

```
2021-03-12 19:07:28 WARNING (Recorder) [homeassistant.components.recorder.migration] Database is about to upgrade. Schema version: 11
2021-03-12 19:07:28 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
```

### mssql

first added the change `ondelete="NO ACTION"` from #46678, then installed the current change

```
2021-03-12 19:03:31 WARNING (Recorder) [homeassistant.components.recorder.migration] Database is about to upgrade. Schema version: 11
2021-03-12 19:03:31 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
```

lmagyar committed:
black formatted

lmagyar commented:
It's ready for review again.

Though I don't know your requirements about test coverage. These changes are tested with real database, that seems to be a stronger validation.

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -206,14 +206,34 @@ def _add_columns(engine, table_name, columns_def):
 def _modify_columns(engine, table_name, columns_def):
     """Modify columns in a table."""
+    if engine.dialect.name == "sqlite":
+        _LOGGER.warning(
```
I don't think this should be debug level since it's only going to generate GitHub issue that nobody will be able to do anything about.  Since SQLite users will already have the short column it won't be a new problem for them so it's not something they need to be warned about.

lmagyar replied:
No, this will never run! We skip sqlite, when we upgrade:
```
if engine.dialect.name != "sqlite":
    _modify_columns(engine, "events", ["event_type VARCHAR(64)"])
```
This is for a later situation, when somebody calls this function for sqlite, and forget, that he shouldn't call this.
But!
Maybe we can skip silently sqlite in `_modify_columns`, because it [doesn't care about column length](https://www.sqlite.org/datatype3.html#affinity_name_examples)? - thanks @B5r1oJ0A9G

lmagyar replied:
I modified it, now we can call `_modify_columns` always, it will skip sqlite silently, only a debug level log!

bdraco reviewed (changes requested):
Looks good. I'll test when I get a chance.  For testing: The ci doesn't have the ability to test anything besides SQLite so the only coverage we could add is to verify it generates the expected sql statements which would be good to have in case someone changes the modify function later and introduces a regression

lmagyar committed:
Skip SQLite altogether.

lmagyar commented:
> the only coverage we could add is to verify it generates the expected sql statements which would be good to have in case someone changes the modify function later and introduces a regression

OK, I will add it soon. Real DB tests can be performed meantime.

bdraco commented:
@lmagyar Let me know if you need help or are stuck on the test as this PR is blocking https://github.com/home-assistant/core/pull/48115

lmagyar commented:
OK, tonight I start to work on it and report on the result next morning!

lmagyar commented:
Sorry, I can't do it. I'm new to HA (I'm coming from Win world), I have no proper dev/test environment, now I tried to set it up, and I think it is impossible on an old Win 7 machine (docker, wsl, etc).

I don't want to delay the merge of this PR, so please somebody take over this PR and write the tests to verify it generates the expected sql statements:
- sqlite (skipped silently)
- mysql (default syntax - `MODIFY {col_def}`)
- postgresql (special handling - `ALTER {column} TYPE {type}`)
- mssql (special handling - `ALTER COLUMN {col_def}`)

raman325 committed:
add test to verify the output of the function

raman325 commented:
I added a test in the linked PR that should allow us to move forward on this

raman325 committed:
fix logic

raman325 committed:
fix docstring

bdraco committed:
fix schema version from merge

bdraco commented:
Testing
- [ ] sqlite (skipped as expected)
- [ ] mariadb
- [ ] postgres

bdraco commented:
postgres looks good

```
2021-04-07 16:24:57 WARNING (Recorder) [homeassistant.components.recorder.migration] Database is about to upgrade. Schema version: 12
2021-04-07 16:24:57 INFO (Recorder) [homeassistant.components.recorder.migration] Upgrading recorder db schema to version 13
2021-04-07 16:24:57 INFO (Recorder) [homeassistant.components.recorder.migration] Upgrade to version 13 done
2021-04-07 16:24:57 INFO (Recorder) [homeassistant.components.recorder.migration] Upgrading recorder db schema to version 14
2021-04-07 16:24:57 WARNING (Recorder) [homeassistant.components.recorder.migration] Modifying columns event_type in table events. Note: this can take several minutes on large databases and slow computers. Please be patient!
2021-04-07 16:24:57 INFO (Recorder) [homeassistant.components.recorder.migration] Upgrade to version 14 done
```

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -206,14 +206,35 @@ def _add_columns(engine, table_name, columns_def):
     """Modify columns in a table."""
+    if engine.dialect.name == "sqlite":
+        _LOGGER.debug(
+            "Skipping to modify columns %s in table %s. "
```
```suggestion
            "Skipping to modify columns %s in table %s; "
```

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -206,14 +206,35 @@ def _add_columns(engine, table_name, columns_def):
+        _LOGGER.debug(
+            "Skipping to modify columns %s in table %s. "
+            "Modifying column length in SQLite is unnecessary, "
+            "it does not impose any length restrictions.",
```
```suggestion
            "it does not impose any length restrictions",
```

Our standard is to not add a period.

bdraco committed:
Update homeassistant/components/recorder/migration.py

bdraco committed:
Update homeassistant/components/recorder/migration.py

bdraco commented on homeassistant/components/recorder/migration.py:
```diff
@@ -206,14 +206,35 @@ def _add_columns(engine, table_name, columns_def):
 def _modify_columns(engine, table_name, columns_def):
     """Modify columns in a table."""
+    if engine.dialect.name == "sqlite":
+        _LOGGER.debug(
```
https://www.sqlite.org/faq.html#q9

bdraco commented:
Verified maria in cli

Verified postgres in UI


<img width="612" alt="Screen Shot 2021-04-07 at 4 31 17 PM" src="https://user-images.githubusercontent.com/663432/113959523-c20acc00-97be-11eb-8ba4-03c7fefa1085.png">

bdraco commented:
I took care of merging everyones work together and fixing the conflicts.

Thanks @Imagyar and @raman325

raman325 commented:
Thanks for the help pushing this through. I'll rebase my other PR and make sure it all works together after the merge

bdraco reviewed (approved):
Testing passed 👍

bdraco commented:
Hi @Imagyar, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#38688).

Once you do that we can merge this pull request.

Thanks!

bdraco reviewed (changes requested):
Should be good to merge once the cla check passes

homeassistant commented:
Hi @lmagyar, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#47748).

Once you do that we will be able to review and accept this pull request.

Thanks!

lmagyar commented:
I'm back, CLA signed! Thank you @bdraco and @raman325 your help!!!

bdraco commented:
Excellent. Let's get this merged.
